bgpd/bgp_route.c

```diff
@@ -2974,25 +2974,52 @@ static int bgp_maximum_prefix_restart_timer(struct thread *thread)
 	return 0;
 }
 
+static uint32_t bgp_filtered_routes_count(struct peer *peer, afi_t afi,
+					  safi_t safi)
+{
+	uint32_t count = 0;
+	struct bgp_dest *dest;
+	struct bgp_adj_in *ain;
+	struct bgp_table *table = peer->bgp->rib[afi][safi];
+
+	for (dest = bgp_table_top(table); dest; dest = bgp_route_next(dest)) {
+		for (ain = dest->adj_in; ain; ain = ain->next) {
+			const struct prefix *rn_p = bgp_dest_get_prefix(dest);
+			struct attr attr = {};
+
+			if (bgp_input_filter(peer, rn_p, &attr, afi, safi)
+			    == FILTER_DENY)
+				count++;
+		}
+	}
+
+	return count;
+}
+
 bool bgp_maximum_prefix_overflow(struct peer *peer, afi_t afi, safi_t safi,
 				 int always)
 {
 	iana_afi_t pkt_afi;
 	iana_safi_t pkt_safi;
+	uint32_t pcount = (CHECK_FLAG(peer->af_flags[afi][safi],
+				      PEER_FLAG_MAX_PREFIX_FORCE))
+				  ? bgp_filtered_routes_count(peer, afi, safi)
+					    + peer->pcount[afi][safi]
+				  : peer->pcount[afi][safi];
```

Comment:
I might be reading this wrong, but it seems like... in 'bgp_filtered_routes_count' only the filtered routes are counted, and not the unfiltered, while 'peer->pcount' only counts unfiltered routes. If so, then this does not sum the two, but uses one or the other, but I only see one or the other here ... ??

Reply:
If _force_ is set then filtered + unfiltered routes are summed, if _force_ is unset, then only unfiltered is used.